### PR TITLE
feat: separate fee discount components

### LINKED ISSUE #2292: Volume discount spec update to better target maker rebate
Following the feedback - I would suggest we either 
a) update the volume discount code / spec to remove maker fee from being subjected to the discount.
b) update the volume discount code / spec so that the volume discount specifically says which of the fee components it’s hitting. So to replicate the current behaviour you’ll have to submit 3 blobs: one for maker fee, one for LP fee and one for infra fee.

### LINKED ISSUE #2296: Add dynamic maker fee based on maker's volume
Dependent on #2295 as the maximum uplift of the maker fee/rebate is the 

Similar to the volume discount scheme, a **new** table of maker fees that maps the volume done by the maker to an extra maker fee factor. Maximum additional fee factor is the total amount allocated to the _network treasury fee factor_ plus the _buyback fee factor_.

This extra amount is added to the maker fee on each trade and is also paid immediately to the price maker.

- [ ] Confirm this should be measuring maker volume only
- [ ] Confirm whether levels are set at a % of the total [maker] volume on that market during the window OR absolute amounts

### PR DESCRIPTION
**Pull request by TomMcL**
Closes #2292

### CODE REVIEW

File: protocol/0029-FEES-fees.md [hunk 1]
```diff
@@ -171,13 +171,21 @@ For example, Ether is 18 decimals (wei). The smallest unit, non divisible is 1 w
     - `infrastructure_referral_fee_discount`
     - `liquidity_fee_referral_discount`
     - `maker_fee_referral_discount`
+1. Referee discounts with differing discounts across the three factors are correctly calculated and applied for each taker fee component during continuous trading (assuming no volume discounts due to party) (<a name="0029-FEES-034" href="#0029-FEES-034">0029-FEES-034</a>)
+    - `infrastructure_referral_fee_discount`
```
Comment (cdummett) [inline protocol/0029-FEES-fees.md]: Me being pedantic so it's consistent everywhere..
```suggestion
    - `infrastructure_fee_referral_discount`
```

File: protocol/0083-RFPR-on_chain_referral_program.md [hunk 2]
```diff
@@ -46,8 +46,14 @@ Enabling or changing the terms of the on-chain referral program can be proposed
 - `benefit_tiers`: a list of dictionaries (with the below fields) defining the reward and discount factors from referrals
   - `minimum_running_notional_taker_volume`: the required [`referral_set_running_notional_taker_volume`](#referral-set-volumes) in quantum units for parties to access this tier
   - `minimum_epochs`: the required number of epochs a referee must have been in a referral set to access this tier
-  - `referral_reward_factor`: the proportion of the referee's taker fees to be rewarded to the referrer
-  - `referral_discount_factor`: the proportion of the referee's taker fees to be discounted
+  - `referral_reward_factors`: the proportions of the referee's taker fees to be rewarded to the referrer
+    - `maker_volume_reward_factor`: the factor to reward the referrer from the maker fee component of any fees paid
+    - `infrastructure_volume_reward_factor`: the factor to reward the referrer from the infrastructure fee component of any fees paid
+    - `lp_volume_reward_factor`: the factor to reward the referrer from the liquidity provider fee component of any fees paid
```
Comment (cdummett) [inline protocol/0083-RFPR-on_chain_referral_program.md]: Slightly prefer the parameter names chosen in `0029` as it makes it clear this reward is coming from the referral program, but have no strong preference other than being consistent (so this suggestion or update `0029` to match the names here).

```suggestion
    - `referral_maker_reward_factor`: the factor to reward the referrer from the maker fee component of any fees paid
    - `referral_infrastructure_reward_factor `: the factor to reward the referrer from the infrastructure fee component of any fees paid
    - `referral_liquidity_reward_factor`: the factor to reward the referrer from the liquidity provider fee component of any fees paid
```

File: protocol/0083-RFPR-on_chain_referral_program.md [hunk 3]
```diff
@@ -46,8 +46,14 @@ Enabling or changing the terms of the on-chain referral program can be proposed
...
+  - `referral_discount_factors`: the proportions of the referee's taker fees to be discounted
+    - `maker_volume_discount_factor`: the factor to discount the maker fee component of any fees paid
+    - `infrastructure_volume_discount_factor`: the factor to discount the infrastructure fee component of any fees paid
+    - `lp_volume_discount_factor`: the factor to discount the liquidity provider fee component of any fees paid
```
Comment (cdummett) [inline protocol/0083-RFPR-on_chain_referral_program.md]: Ditto above comment.

```suggestion
    - `referral_maker_discount_factor `: the factor to discount the maker fee component of any fees paid
    - `referral_infrastructure_discount_factor `: the factor to discount the infrastructure fee component of any fees paid
    - `referral_liquidity_discount_factor `: the factor to discount the liquidity provider fee component of any fees paid
```

File: protocol/0083-RFPR-on_chain_referral_program.md [hunk 4]
```diff
@@ -61,20 +67,44 @@ message UpdateReferralProgram{
...
             {
                 "minimum_running_notional_taker_volume": 30000,
                 "minimum_epochs": 31,
-                "referral_reward_factor": 0.010,
-                "referral_discount_factor": 0.010,
+                "referral_reward_factors": {
+                    "maker_volume_discount_factor": 0.01,
+                    "infrastructure_volume_discount_factor": 0.01,
+                    "lp_volume_discount_factor": 0.01
+                },
```
Comment (cdummett) [inline protocol/0083-RFPR-on_chain_referral_program.md]: Should be `reward` not `discount` here and match the names chosen in `0029`.

File: protocol/0083-RFPR-on_chain_referral_program.md [hunk 5]
```diff
@@ -329,20 +359,44 @@ Given:
         {
             "minimum_running_notional_taker_volume": 10000,
             "minimum_epochs": 0,
-            "referral_reward_factor": 0.001,
-            "referral_discount_factor": 0.001,
+            "referral_reward_factor": {
+                "maker_volume_discount_factor": 0.001,
+                "infrastructure_volume_discount_factor": 0.001,
+                "lp_volume_discount_factor": 0.001
```
Comment (cdummett) [inline protocol/0083-RFPR-on_chain_referral_program.md]: Ditto here, `reward` not `discount` and match `0029`.

File: protocol/0083-RFPR-on_chain_referral_program.md [hunk 6]
```diff
@@ -329,20 +359,44 @@ Given:
...
         },
         {
             "minimum_running_notional_taker_volume": 20000,
             "minimum_epochs": 7,
-            "referral_reward_factor": 0.005,
-            "referral_discount_factor": 0.005,
+            "referral_reward_factor": {
+                "maker_volume_discount_factor": 0.005,
+                "infrastructure_volume_discount_factor": 0.005,
+                "lp_volume_discount_factor": 0.005
```
Comment (cdummett) [inline protocol/0083-RFPR-on_chain_referral_program.md]: Ditto here, `reward` not `discount` and match `0029`.

File: protocol/0083-RFPR-on_chain_referral_program.md [hunk 7]
```diff
@@ -329,20 +359,44 @@ Given:
...
         },
         {
             "minimum_running_notional_taker_volume": 30000,
             "minimum_epochs": 31,
-            "referral_reward_factor": 0.010,
-            "referral_discount_factor": 0.010,
+            "referral_reward_factor": {
+                "maker_volume_discount_factor": 0.01,
+                "infrastructure_volume_discount_factor": 0.01,
+                "lp_volume_discount_factor": 0. 01
```
Comment (cdummett) [inline protocol/0083-RFPR-on_chain_referral_program.md]: Ditto here, `reward` not `discount` and match `0029`.

And should be `referral` not `volume`.

File: protocol/0083-RFPR-on_chain_referral_program.md [hunk 8]
```diff
@@ -329,20 +359,44 @@ Given:
...
+            },
+            "referral_discount_factor": {
+                "maker_volume_discount_factor": 0.01,
+                "infrastructure_volume_discount_factor": 0.01,
+                "lp_volume_discount_factor": 0.01
```
Comment (cdummett) [inline protocol/0083-RFPR-on_chain_referral_program.md]: This one should be `referral` not `volume`?

File: protocol/0084-VDPR-volume_discount_program.md [hunk 9]
```diff
@@ -17,7 +17,10 @@ Enabling or changing the terms of the volume discount program can be proposed vi
 
 - `benefit_tiers`: a list of dictionaries with the following fields
   - `minimum_party_running_notional_taker_volume`: the required `party_running_notional_taker_volume` in quantum units for a party to access this tier
-  - `volume_discount_factor`: the proportion of the referees taker fees to be rewarded to the referrer
+  - `volume_discount_factors`: a dictionary with the following fields
+    - `maker_volume_discount_factor`: the factor to discount the maker fee component of any fees paid
+    - `infrastructure_volume_discount_factor`: the factor to discount the infrastructure fee component of any fees paid
+    - `lp_volume_discount_factor`: the factor to discount the liquidity provider fee component of any fees paid
```
Comment (cdummett) [inline protocol/0084-VDPR-volume_discount_program.md]: Again be consistent with `0029`.

**Review by cdummett — CHANGES_REQUESTED** [review]
Mostly nitpicks about consistency of parameter names.

The formulas and mechanics seem good to me though 👍

File: protocol/0029-FEES-fees.md [hunk 10]
```diff
@@ -82,10 +86,14 @@ Note, discounts are calculated and applied one after the other and **before** re
 - infrastructure: staking/governance system/engine (network wide)
 - maker: market framework / market making (network wide)
 - liquidity: market making system (per market)
+- treasury: Fees sent to network treasury for later usage via governance votes (network wide)
+- buyback: Fees used to purchase governance tokens to the protocol via regular auctions (network wide)
 
 The infrastructure fee factor is set by a network parameter `market.fee.factors.infrastructureFee` and a reasonable default value is `fee_factor[infrastructure] = 0.0005 = 0.05%`.
 The maker fee factor is set by a network parameter `market.fee.factors.makerFee` and a reasonable default value is `fee_factor[maker] = 0.00025 = 0.025%`.
 The liquidity fee factor is set by an auction-like mechanism based on the liquidity provisions committed to the market, see [setting LP fees](./0042-LIQF-setting_fees_and_rewarding_lps.md).
+The treasury fee factor is set by the network parameter `market.fee.factors.treasuryFee`.
```
Comment (ze97286) [inline protocol/0029-FEES-fees.md]: valid/default values?

File: protocol/0029-FEES-fees.md [hunk 11]
```diff
@@ -107,15 +115,25 @@ If the transfer fails:
...
 
-The transfer of fees must be completed before performing the normal post-trade calculations (MTM Settlement, position resolution etc...). The transfers have to be identifiable as fee transfers and separate for the three components.
+Additionally, a `high_volume_market_maker_rebate` may be necessary which will be taken from the `treasury/buyback_fee` components. This will be calculated as:
+1. Determine whether the maker party of the trade (if there is one) qualifies for a `high volume market maker rebate` and, if so, at what rate.
+1. Calculate the fee that this corresponds to as `high_volume_maker_fee = high_volume_factor * trade_value_for_fee_purposes`.
+1. Take this fee from the `treasury_fee` and `buyback_fee` (protocol restrictions on governance changes ensure that `treasury_fee + buyback_fee >= high_volume_maker_fee` is always true) as a proportion of their relative sizes, i.e.:
+   1. `high_volume_maker_fee = high_volume_factor * trade_value_for_fee_purposes`
+   1. `treasury_fee = treasury_fee * (1 - high_volume_maker_fee / (treasury_fee + buyback_fee))`
+   1. `buyback_fee = treasury_fee * (1 - buyback_fee / (treasury_fee + buyback_fee))`
```
Comment (witgaw) [inline protocol/0029-FEES-fees.md]: Shouldn't it be?
```suggestion
   1. `treasury_fee = treasury_fee * (1 - high_volume_maker_fee * treasury_fee / (treasury_fee + buyback_fee))`
   1. `buyback_fee = buyback_fee * (1 - high_volume_maker_fee * buyback_fee / (treasury_fee + buyback_fee))`
```

File: protocol/0095-HVMR-high_volume_maker_rebate.md [hunk 12]
```diff
@@ -0,0 +1,26 @@
...
+- `end_of_program_timestamp`: the timestamp after which when the current epoch ends, the program will become inactive and benefits will be disabled. If this field is empty, the program runs indefinitely.
+- `window_length`:  the number of epochs over which to average a party's `maker_volume_fraction`
+
+## Application
+
+As variable fees for the taker depending upon with whom they are trading would not be a good experience, the additional maker rebate should be taken from a weighted combination of the network treasury and network buyback components of the total fee. The exact calculations are laid out in [0029-FEES](./0029-FEES-fees.md) but are broadly:
+
+   1. `high_volume_maker_fee = high_volume_factor * trade_value_for_fee_purposes`
+   1. `treasury_fee = treasury_fee * (1 - high_volume_maker_fee / (treasury_fee + buyback_fee))`
+   1. `buyback_fee = treasury_fee * (1 - buyback_fee / (treasury_fee + buyback_fee))`
```
Comment (witgaw) [inline protocol/0095-HVMR-high_volume_maker_rebate.md:22]:
```suggestion
   1. `treasury_fee = treasury_fee * (1 - high_volume_maker_fee * treasury_fee / (treasury_fee + buyback_fee))`
   1. `buyback_fee = buyback_fee * (1 - high_volume_maker_fee * buyback_fee / (treasury_fee + buyback_fee))`
```

**Review by witgaw — COMMENTED** [review]
LGTM, just unsure about the formula.

File: protocol/0029-FEES-fees.md (hunk 11 above)
Comment (TomMcL) [reply protocol/0029-FEES-fees.md]: Have changed these up now, not quite what is in this suggested change but *I think* correct now so please double check :D

Comment (witgaw) [reply protocol/0029-FEES-fees.md]: Goood point, looks good now 👍

File: protocol/0029-FEES-fees.md (hunk 10 above)
Comment (witgaw) [reply protocol/0029-FEES-fees.md]: Added `0` for both

**Review by cdummett — APPROVED** [review]

**Pull request merged**